### PR TITLE
fix: assets ending in ".$OS.$ARCH"

### PR DESCRIPTION
**Pull request by jdx**
Fixes https://github.com/jdx/mise/issues/2854

### CODE REVIEW

**Comment by autarch** [comment]
Hi, thanks for your PR! I'm pretty finicky about my projects (see [this blog post](https://blog.urth.org/2022/07/02/what-s-the-right-way-to-merge-a-pull-request/) for details), so I rarely merge a PR as-is. I can move forward on your PR in one of two ways:

1. I check it out locally, fiddle with it as needed, merge it locally, and simply close this PR. This will preserve at least one commit with your name on it, but the PR will show up as closed in your GitHub stats.
2. If you enable me to push directly to your fork, I can do my fiddling, then force push to your fork and merge the resulting PR. Again, this will preserve at least one commit with your name on it, but you _also_ get credit for the PR merge in your GitHub stats. The only downside is that I will be force pushing directly to your fork.

Please let me know which approach you'd prefer. If I don't hear from you before I get around to working on this PR I'll go with option #1.

Thanks again for your contribution!

File: ubi/src/os.rs [hunk 1]
```diff
@@ -52,3 +52,29 @@ pub(crate) fn all_oses_re() -> Regex {
...
+            regex!(r"(?i:(?:\b|_)(?:mips|mips64)(?:\b|_))"),
+            regex!(r"(?i:(?:\b|_)(?:ppc|ppc64)(?:\b|_))"),
+            regex!(r"(?i:(?:\b|_)(?:s390x)(?:\b|_))"),
+        ]
+        .iter()
+        .map(|r| format!("({})", r.as_str()))
+        .join("|"),
+    )
+    .unwrap()
+}
```
Comment (autarch) [inline ubi/src/os.rs:70]: This already exists in the `arch` module.

File: ubi/src/os.rs [hunk 2]
```diff
@@ -52,3 +52,29 @@ pub(crate) fn all_oses_re() -> Regex {
...
+
+pub(crate) static PLATFORM_RE: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(
+        &[all_oses_re(), all_arches_re()]
+            .iter()
+            .map(|r| format!("({})", r.as_str()))
+            .join("|"),
+    )
+    .unwrap()
+});
```
Comment (autarch) [inline ubi/src/os.rs:80]: I think this can be constructed in the `extension` module by importing the relevant fns from `os` and `arch`.

**Review by autarch — CHANGES_REQUESTED** [review]
Thanks for this fix. This looks good % some small stuff. I can make the changes myself or you can do it if you'd prefer. Please let me know.

File: ubi/src/os.rs (hunk 2 above)
Comment (jdx) [reply ubi/src/os.rs:80]: idk how significant it is but won't that involve recompiling the final regex many times?

Comment (autarch) [reply ubi/src/os.rs]: I thought the use of `Lazy` would mean it's only compiled once, right?

Comment (autarch) [reply ubi/src/os.rs]: Or rather, can't `Lazy` also be used in the `extension` module? Maybe there's some issue with doing that?

Comment (jdx) [reply ubi/src/os.rs]: all_oses_re is not wrapped in Lazy and gets called many times

Comment (autarch) [reply ubi/src/os.rs]: Ah, right. I just pushed a change to make both of these `static Lazy<Regex>` vars instead of functions.

**Comment by jdx** [comment]
rebased

File: ubi/src/extension.rs [hunk 3]
```diff
@@ -108,7 +110,17 @@ fn extension_is_part_of_version(path: &str, ext_str: &OsStr) -> bool {
 }
 
 fn extension_is_platform(ext_str: &OsStr) -> bool {
-    ALL_OSES_RE.is_match(ext_str.to_string_lossy().as_ref())
+    static PLATFORM_RE: Lazy<Regex> = Lazy::new(|| {
+        Regex::new(
+            &[ALL_OSES_RE.as_str(), ALL_ARCHES_RE.as_str()]
+                .iter()
+                .map(|r| format!("({r})"))
```
Comment (autarch) [inline ubi/src/extension.rs:117]:
```suggestion
                .map(|r| format!("(?:{r})"))
```

Comment (autarch) [reply ubi/src/extension.rs:117]: I'm going to add this to the os and arch regexes too. This PR just made me notice this was missing.

**Review by autarch — CHANGES_REQUESTED** [review]
LGTM % one very small thing.

**Comment by autarch** [comment]
Is it okay if I push to this PR. I want to tweak the commit message before merging.

**Comment by jdx** [comment]
yeah that's fine

**Review by autarch — APPROVED** [review]

**Pull request merged**